File: samples/java/src/main/java/Endorser.java
import org.hyperledger.indy.sdk.anoncreds.AnoncredsResults;
import org.hyperledger.indy.sdk.pool.Pool;
import org.hyperledger.indy.sdk.wallet.Wallet;
import org.hyperledger.indy.sdk.did.Did;
import org.hyperledger.indy.sdk.did.DidJSONParameters;
import org.hyperledger.indy.sdk.did.DidResults.CreateAndStoreMyDidResult;
import org.json.JSONArray;
import org.json.JSONObject;
import utils.PoolUtils;

import java.util.Arrays;

import static org.hyperledger.indy.sdk.ledger.Ledger.buildNymRequest;
import static org.hyperledger.indy.sdk.ledger.Ledger.buildSchemaRequest;
import static org.hyperledger.indy.sdk.ledger.Ledger.signAndSubmitRequest;
import static org.hyperledger.indy.sdk.ledger.Ledger.submitRequest;
import static org.hyperledger.indy.sdk.ledger.Ledger.multiSignRequest;
import static org.hyperledger.indy.sdk.ledger.Ledger.appendRequestEndorser;
import static org.hyperledger.indy.sdk.anoncreds.Anoncreds.issuerCreateSchema;
import static org.junit.Assert.assertEquals;

class Endorser {
    static void demo() throws Exception {

        System.out.println("Endorser sample -> started");
        String trusteeSeed = "000000000000000000000000Trustee1";

        // Set protocol version 2 to work with Indy Node 1.4
        Pool.setProtocolVersion(PoolUtils.PROTOCOL_VERSION).get();

        // 1. Create and Open Pool
        String poolName = PoolUtils.createPoolLedgerConfig();
        Pool pool = Pool.openPoolLedger(poolName, "{}").get();

        // 2. Create and Open Author Wallet
        String authorWalletConfig = new JSONObject().put("id", "authorWallet").toString();
        String authorWalletCredentials = new JSONObject().put("key", "author_wallet_key").toString();
        Wallet.createWallet(authorWalletConfig, authorWalletCredentials).get();
        Wallet authorWallet = Wallet.openWallet(authorWalletConfig, authorWalletCredentials).get();

        // 3. Create and Open Endorser Wallet
        String endorserWalletConfig = new JSONObject().put("id", "endorserWallet").toString();
        String endorserWalletCredentials = new JSONObject().put("key", "endorser_wallet_key").toString();
        Wallet.createWallet(endorserWalletConfig, endorserWalletCredentials).get();
        Wallet endorserWallet = Wallet.openWallet(endorserWalletConfig, endorserWalletCredentials).get();

        // 3. Create and Open Trustee Wallet
        String trusteeWalletConfig = new JSONObject().put("id", "trusteeWallet").toString();
        String trusteeWalletCredentials = new JSONObject().put("key", "trustee_wallet_key").toString();
        Wallet.createWallet(trusteeWalletConfig, trusteeWalletCredentials).get();
        Wallet trusteeWallet = Wallet.openWallet(trusteeWalletConfig, trusteeWalletCredentials).get();

        // 4. Create Trustee DID
        DidJSONParameters.CreateAndStoreMyDidJSONParameter theirDidJson =
                new DidJSONParameters.CreateAndStoreMyDidJSONParameter(null, trusteeSeed, null, null);
        CreateAndStoreMyDidResult createTheirDidResult = Did.createAndStoreMyDid(trusteeWallet, theirDidJson.toJson()).get();
        String trusteeDid = createTheirDidResult.getDid();

        // 5. Create Author DID
        CreateAndStoreMyDidResult createMyDidResult = Did.createAndStoreMyDid(authorWallet, "{}").get();
        String authorDid = createMyDidResult.getDid();
        String authorVerkey = createMyDidResult.getVerkey();

        // 6. Create Endorser DID
        createMyDidResult = Did.createAndStoreMyDid(endorserWallet, "{}").get();
        String endorserDid = createMyDidResult.getDid();
        String endorserVerkey = createMyDidResult.getVerkey();

        // 7. Build Author Nym Request
        String nymRequest = buildNymRequest(trusteeDid, authorDid, authorVerkey, null, null).get();

        // 8. Trustee Sign Author Nym Request
        signAndSubmitRequest(pool, trusteeWallet, trusteeDid, nymRequest).get();

        // 9. Build Endorser Nym Request
        nymRequest = buildNymRequest(trusteeDid, endorserDid, endorserVerkey, null, "ENDORSER").get();

        // 10. Trustee Sign Endorser Nym Request
        signAndSubmitRequest(pool, trusteeWallet, trusteeDid, nymRequest).get();

        // 11. Create schema with endorser

        String schemaName = "gvt";
        String schemaVersion = "1.0";
        String schemaAttributes = new JSONArray().put("name").put("age").put("sex").put("height").toString();
        AnoncredsResults.IssuerCreateSchemaResult createSchemaResult =
                issuerCreateSchema(authorDid, schemaName, schemaVersion, schemaAttributes).get();
        String schemaId = createSchemaResult.getSchemaId();
        String schemaJson = createSchemaResult.getSchemaJson();

        //  Transaction Author builds Schema Request
        String schemaRequest = buildSchemaRequest(authorDid, schemaJson).get();

        //  Transaction Author appends Endorser's DID into the request
        String schemaRequestWithEndorser = appendRequestEndorser(schemaRequest, endorserDid).get();

        //  Transaction Author signs the request with the added endorser field
        String schemaRequestWithEndorserAuthorSigned =
                multiSignRequest(authorWallet, authorDid, schemaRequestWithEndorser).get();

        //  Transaction Endorser signs the request
        String schemaRequestWithEndorserSigned =
                multiSignRequest(endorserWallet, endorserDid, schemaRequestWithEndorserAuthorSigned).get();

        //  Transaction Endorser sends the request
        String response = submitRequest(pool, schemaRequestWithEndorserSigned).get();
        JSONObject responseJson = new JSONObject(response);
        assertEquals("REPLY", responseJson.getString("op"));

        pool.closePoolLedger().get();
        Pool.deletePoolLedgerConfig(poolName).get();

        trusteeWallet.closeWallet().get();
        Wallet.deleteWallet(trusteeWalletConfig, trusteeWalletCredentials).get();

        authorWallet.closeWallet().get();
        Wallet.deleteWallet(authorWalletConfig, authorWalletCredentials).get();

        endorserWallet.closeWallet().get();
        Wallet.deleteWallet(endorserWalletConfig, endorserWalletCredentials).get();

        System.out.println("Endorser sample -> completed");
    }
}

File: samples/python/src/endorser.py
import time
import json
import logging

from src.utils import get_pool_genesis_txn_path, run_coroutine, PROTOCOL_VERSION


logger = logging.getLogger(__name__)

from indy import ledger, did, wallet, pool, anoncreds


async def open_pool(target_pool):
    target_pool['genesis_txn_path'] = get_pool_genesis_txn_path(target_pool['pool_name'])
    target_pool['pool_config'] = json.dumps({"genesis_txn": str(target_pool['genesis_txn_path'])})
    await pool.create_pool_ledger_config(target_pool['pool_name'], target_pool['pool_config'])
    target_pool['pool'] = await pool.open_pool_ledger(target_pool['pool_name'], None)


async def setup_identity(identity, trustee):
    (identity['did'], identity['verkey']) = await did.create_and_store_my_did(identity['wallet'], "{}")
    nym_req = await ledger.build_nym_request(trustee['did'], identity['did'], identity['verkey'], None,
                                             identity['role'])
    await ledger.sign_and_submit_request(identity['pool'], trustee['wallet'], trustee['did'], nym_req)


async def demo():
    logger.info("endorser sample -> started")

    # Set protocol version 2 to work with Indy Node 1.4
    await pool.set_protocol_version(PROTOCOL_VERSION)

    # Open pool
    target_pool = {
        'pool_name': 'trustee_pool'
    }

    await open_pool(target_pool)

    author = {
        'wallet_config': json.dumps({'id': 'author_wallet'}),
        'wallet_credentials': json.dumps({'key': 'author_wallet_key'}),
        'pool': target_pool['pool'],
        'role': None
    }
    endorser = {
        'wallet_config': json.dumps({"id": "endorser_wallet"}),
        'wallet_credentials': json.dumps({"key": "endorser_wallet_key"}),
        'pool': target_pool['pool'],
        'role': 'ENDORSER'
    }
    trustee = {
        'seed': '000000000000000000000000Trustee1',
        'wallet_config': json.dumps({'id': 'trustee_wallet'}),
        'wallet_credentials': json.dumps({'key': 'trustee_wallet_key'}),
        'pool': target_pool['pool']
    }

    await wallet.create_wallet(author['wallet_config'], author['wallet_credentials'])
    author['wallet'] = await wallet.open_wallet(author['wallet_config'], author['wallet_credentials'])

    await wallet.create_wallet(endorser['wallet_config'], endorser['wallet_credentials'])
    endorser['wallet'] = await wallet.open_wallet(endorser['wallet_config'], endorser['wallet_credentials'])

    await wallet.create_wallet(trustee['wallet_config'], trustee['wallet_credentials'])
    trustee['wallet'] = await wallet.open_wallet(trustee['wallet_config'], trustee['wallet_credentials'])

    # Trustee Create DID
    (trustee['did'], trustee['verkey']) = \
        await did.create_and_store_my_did(trustee['wallet'], json.dumps({"seed": trustee['seed']}))

    # Author Create DID
    await setup_identity(author, trustee)

    #  Endorser Create DID
    await setup_identity(endorser, trustee)

    # Write SCHEMA from author with endorser
    # Endorser Workflow
    schema = {
        'name': 'gvt',
        'version': '1.0',
        'attributes': '["age", "sex", "height", "name"]'
    }
    author['schema_id'], author['schema'] = await anoncreds.issuer_create_schema(author['did'], schema['name'],
                                                                                 schema['version'],
                                                                                 schema['attributes'])

    # 1. Transaction Author builds Schema Request
    author['schema_req'] = await ledger.build_schema_request(author['did'], author['schema'])

    # 2. Transaction Author appends Endorser's DID into the request
    author['schema_req'] = await ledger.append_request_endorser(author['schema_req'], endorser['did'])

    # 3. Transaction Author signs the request with the added endorser field
    author['schema_req'] = await ledger.multi_sign_request(author['wallet'], author['did'], author['schema_req'])

    # 4. Transaction Author sends the request to the Endorser (out of scope).
    endorser['schema_req'] = author['schema_req']

    # 5. Transaction Endorser signs the request
    endorser['schema_req'] = await ledger.multi_sign_request(endorser['wallet'], endorser['did'],
                                                             endorser['schema_req'])

    # 6. Transaction Endorser sends the request
    resp = await ledger.submit_request(target_pool['pool'], endorser['schema_req'])
    resp = json.loads(resp)
    assert resp['op'] == 'REPLY'
    assert resp['result']['txn']['metadata']['endorser'] == endorser['did']

    time.sleep(1)

    # Author checks that Schema was added
    author['get_schema_req'] = await ledger.build_get_schema_request(author['did'], author['schema_id'])
    response = await ledger.submit_request(target_pool['pool'], author['get_schema_req'])
    response = json.loads(response)
    assert response['op'] == 'REPLY'

    await pool.close_pool_ledger(target_pool['pool'])
    await pool.delete_pool_ledger_config(target_pool['pool_name'])

    await wallet.close_wallet(trustee['wallet'])
    await wallet.delete_wallet(trustee['wallet_config'], trustee['wallet_credentials'])

    await wallet.close_wallet(author['wallet'])
    await wallet.delete_wallet(author['wallet_config'], author['wallet_credentials'])

    await wallet.close_wallet(endorser['wallet'])
    await wallet.delete_wallet(endorser['wallet_config'], endorser['wallet_credentials'])

    logger.info("endorser sample -> completed")


if __name__ == '__main__':
    run_coroutine(demo)
    time.sleep(1)  # FIXME waiting for libindy thread complete
